# Metagenomic Analysis on Ancient Samples

## Introduction

In Day 2 and Day 3, we will perform metagenomic analysis on ancient samples which are generated by Hannesâ€™s group ( unpublished ). It mainly contains three parts. 

1. Perform the metagenomic screening using Metaphlan4
2. Validate and authenticate the ancient genomes
3. Visualize the ancient genomes

After the exercise, you will be able to: 
* Identify the metagenomic profiles/taxonomic composition within the ancient sample 
* Authenticate the target genomes based on damage pattern/edit distance/evenness coverage, etc
* Generate the plot for the ancient metagenomic analysis

## Metagenomic screening

1. Load the working environment and make a folder for the results 
 ```
 mkdir metaphlan
 cd metaphlan
 cp /projects/course_sgbb20001/people/hsf378/metaphlan/Metaphlan.sh ./ 
 conda activate /projects/course_sgbb20001/data/envs/metaphlan4
 sbatch Metaphlan.sh
 ```
2. Run metagenomic screening on Metaphlan
 
- if you only have one sample, eg. Lola
 ```
 metaphlan  /projects/course_sgbb20001/people/hsf378/Eager/results/samtools/filter/Lola.unmapped.fastq.gz \
  --nproc 16 \
  --read_min_len 30 \
  --input_type fastq \
  --bowtie2out Lola.bowtie2.out \
  -o Lola.metaphlan4.txt \
  --bowtie2db /datasets/globe_databases/metaphlan/20240619 
 ```
 By default, metaphlan perform the metagenomic profiling on bacteria, archaea and eukaryotes. If you want to include the viruses, you have to **change the bowtie2 database which set up for Metaphlan 3** and add extra parameters `--mpa3 --add_viruses`. 

 * Check the output files

<details><summary>Extend to check files within the folder</summary>

>   Lola.bowtie2.out <br/>
>   Lola.metaphlan4.txt  <br/>

</details>

 * Check the bowtie2.out

<details><summary>Click</summary>

 | header of reads | marker gene |
 | :--- | :--- |
 | M_A00706:484:H25WWDSX3:2:1451:20401:33599_RG:Z:ILLUMINA-MSW-2765_SCR1_HXZD2 | 194702__GCA_900460955.1_00324 |
 | M_A00706:484:H25WWDSX3:2:2554:4698:1297_RG:Z:ILLUMINA-MSW-2765_SCR1_HXZD2 | 563038__E9FKL8__HMPREF0851_00827 |
 | M_A00706:484:H25WWDSX3:2:2519:31412:28025_RG:Z:ILLUMINA-MSW-2765_SCR1_HXZD2 |999425__K8MRR9__HMPREF9186_01101 |
 | M_A00706:484:H25WWDSX3:2:2619:28140:6605_RG:Z:ILLUMINA-MSW-2765_SCR1_HXZD2 |544581__G9PQL4__HMPREF1979_01803 |
 | M_A00706:484:H25WWDSX3:2:2166:3170:31532_RG:Z:ILLUMINA-MSW-2765_SCR1_HXZD2 |43675__D2NSN9__RM6536_1480 |
 | M_A00706:484:H25WWDSX3:2:2173:24080:36401_RG:Z:ILLUMINA-MSW-2765_SCR1_HXZD2 |2047__E3H1D8__HMPREF0733_10607 |
 | M_A00706:484:H25WWDSX3:2:2224:10366:15060_RG:Z:ILLUMINA-MSW-2765_SCR1_HXZD2 |487__A0A0Y5FL51__galM |
 | M_A00706:484:H25WWDSX3:2:2532:18945:21371_RG:Z:ILLUMINA-MSW-2765_SCR1_HXZD2 |487__A0A378WD25__NM65014_0252 |
 | M_A00706:357:HKNKTDRXY:2:2101:32633:5118_RG:Z:ILLUMINA-MSW-2765_SCR1_4Y2F7 |1263109__GCA_000438135.1_00036 |
 | M_A00706:484:H25WWDSX3:2:2158:4083:8030_RG:Z:ILLUMINA-MSW-2765_SCR1_HXZD2 |712121__L1PF62__HMPREF9061_01305 |

</details>

 * Check the metaphlan4.txt

<details><summary>Click</summary>

#mpa_v30_CHOCOPhlAn_201901 <br/>
#/projects/mjolnir1/apps/conda/metaphlan-3.0.14/bin/metaphlan /projects/mjolnir1/people/hsf378/data/Metagenomic_workshop/raw/MSW-2765.fq.gz --nproc 16 --read_min_len 30 --input_type fastq --bowtie2out MSW-2765.bowtie2.out -o MSW-2765.metaphlan4.txt --bowtie2db /home/jql545/data/FromJonas/Bowtie2DB --mpa3 --add_viruses <br/>
#SampleID	Metaphlan_Analysis <br/>
 |#clade_name | NCBI_tax_id | relative_abundance | additional_species | 
 | --- | --- | --- | --- | 
 | k__Bacteria | 2 | 81.89136 |  |
 | k__Viruses | 10239 | 18.03597 |  | 
 | k__Archaea | 2157 | 0.07268 |  | 
 | k__Bacteria\|p__Firmicutes | 2\|1239 | 30.84009 |  | 
 | k__Bacteria\|p__Proteobacteria | 2\|1224 | 26.65484 |  | 
 | k__Viruses\|p__Viruses_unclassified | 10239\| | 18.03597 |  | 

</details>

3. Merge all output tables
 
 ```
 merge_metaphlan_tables.py Lola.metaphlan4.txt /projects/course_sgbb20001/people/hsf378/metaphlan/modern_ref/*metaphlan4.txt > merged_abundance_table.txt
 ```

 * Check the merged abundance file
  ```
  head merge.metaphlan4.txt
  ```
 * Modify the output for the downstream visualization. 
 ```
 grep -E "(^ID)|(g__)" merged_abundance_table.txt | grep -v "s__" | sed "s/.*g__//g" > merged_genera.txt
 sed -n 2p merged_abundance_table.txt | cat - merged_genera.txt  | sed 's/.metaphlan4//g' > MetaPhlan_GeneraComparison.tsv
 ```
 * Check the modified abundance file
 ```
 head MetaPhlan_GeneraComparison.tsv
 ```

 <details><summary>Click</summary>

 | clade_name | MSW-2765 | HOR-1 | GIL-1 |
 | --- | --- | --- | --- |
 | Candidatus_Bathyarchaeota_archaeon_CG_4_8_14_3_um_filter_42_8 | 0.02647 |0 | 0 |
 | Candidatus_Bathyarchaeota_archaeon_RBG_13_46_16b | 0 | 0 | 0.01343 | 
 | Candidatus_Lokiarchaeota_archaeon_CR_4 | 0.0462 | 0.27318 | 0.11878 | 
 | Pyrinomonas_methylaliphatogenes | 0.08442 | 0 | 0 | 
 | Actinobaculum_sp_oral_taxon_183 | 0.57967 | 0.10377 | 0 | 
 | Actinomyces_georgiae | 0.00688 | 0 | 0 | 
 | Actinomyces_graevenitzii | 0 | 0.25428 | 0 | 
 | Actinomyces_johnsonii | 0.11789 | 0.05306 | 0 | 
 | Actinomyces_meyeri | 0.09658 | 0 | 0 | 

 </details>

4. Visualize the abundance profiles

 Here, hclust2 is applied to generate the heatmap plot for the merged metaphlan abundance profiles. 

 ```
 module load hclust2

 hclust2.py -i MetaPhlan_GeneraComparison.tsv -o MetaPhlanTop30Genera.png --f_dist_f braycurtis --s_dist_f braycurtis --cell_aspect_ratio 0.5 -l --flabel_size 10 --slabel_size 10 --max_flabel_len 100 --max_slabel_len 100 --dpi 1000 --ftop 30

 ```
 
* Check the heatmap plot
 
 ```
 scp /projects/course_sgbb20001/people/hsf378/metaphlan/MetaPhlanTop30Genera.png ./

 ```

The top 30 abundance Genera bacteria/archaea hists on Lola sample, as the following figure. <img width="1020" alt="Top30" src="https://github.com/Schroeder-Group/aDNA-course/MetaPhlanTop30Genera.png">

## Validation and Authentication

1. Align reads to the GTDB database including representative reference genomes of bacteria, archaea, viruses and eukaryotes.

```
cd ../
mkdir -p Bowtie2
cd Bowtie2
cp /projects/course_sgbb20001/people/hsf378/Bowtie2/bowtie2.sh ./
module load bowtie2/2.5.2 samtools/1.17
sbatch bowtie2.sh
```

```
FASTQ_DIRECTORY=/projects/course_sgbb20001/people/hsf378/Eager/results/samtools/filter
SAMPLE=Lola
DB=/projects/course_sgbb20001/data/databases/Vanilla/gtdb-r202-organelles-viruses

bowtie2 -x $DB -U $FASTQ_DIRECTORY/$SAMPLE.unmapped.fastq.gz -p 10 --maxins 500 \
        --rg-id ILLUMINA-$SAMPLE --rg SM:$SAMPLE \
        --rg PL:illumina --rg PU:ILLUMINA-$SAMPLE -k 16 --np 1 \
        --mp "1,1" --rdg "0,1" --rfg "0,1" --score-min "L,0,-0.05" \
        --very-sensitive --no-unal  | samtools view -@ 10 -b -F 4 - | samtools sort -@ 10 -O bam - > $SAMPLE.vanilla.sorted.bam
```

> Here the parameter `-k 16`: maximum 16 alignments per reads.
> `--maxins 500`: the maximum of reads length is 500bp.
> `--rg-id ILLUMINA-$SAMPLE --rg SM:$SAMPLE --rg PL:illumina --rg PU:ILLUMINA-$SAMPLE`: set the flag for header.
> `--np 1 --mp "1,1" --rdg "0,1" --rfg "0,1" --score-min "L,0,-0.05"`: specify how bowtie2 scores the alignments.

* Check the alignment results:

> 7827212 reads; of these: <br/>
>  7827212 (100.00%) were unpaired; of these: <br/>
>    661246 (8.45%) aligned 0 times <br/>
>    2446017 (31.25%) aligned exactly 1 time <br/>
>    4719949 (60.30%) aligned >1 times <br/>
> 91.55% overall alignment rate <br/>

After the mapping steps, we will remove the duplicates reads with samtools rmdup.

```
samtools rmdup -s $SAMPLE.vanilla.sorted.bam $SAMPLE.vanilla.rmdup.bam
samtools index $SAMPLE.vanilla.rmdup.bam
```

2. Filter out the noise from multiple aligned reads, and those reference genomes with uneven coverage by using FilterBAM.

```
cd ../
mkdir -p filterbam
cd filterbam
cp /projects/course_sgbb20001/people/hsf378/filterbam/filterbam.sh ./
conda activate /projects/course_sgbb20001/data/envs/bam-filterV2/
sbatch filterbam.sh
```

<details><summary>Click</summary>
 
 ```
 BAM_DIRECTORY=/projects/course_sgbb20001/people/hsf378/Bowtie2
 SAMPLE=Lola

 filterBAM reassign --bam $BAM_DIRECTORY/$SAMPLE.vanilla.rmdup.bam  --threads 10 \
        --iters 0 --min-read-ani 92 \
        --reference-lengths /projects/course_sgbb20001/data/databases/Vanilla/gtdb-r202-organelles-viruses.len.map \
        --out-bam $SAMPLE.reassigned.bam
      
 filterBAM filter --bam $SAMPLE.reassigned.bam --bam-filtered $SAMPLE.dedup.filtered.bam \
        --stats $SAMPLE.dedup.stats.tsv.gz --stats-filtered $SAMPLE.dedup.stats-filtered.tsv.gz --threads 10 --min-read-ani 92 --min-normalized-entropy 0.6
 ```

</details>

For example on the GIL-1 sample:

> INFO ::: 2022-12-07 05:37:49 ::: Loading BAM file  <br/>
> INFO ::: 2022-12-07 05:37:51 ::: BAM index not found. Indexing... <br/>
> INFO ::: 2022-12-07 05:37:55 ::: Reloading BAM file <br/>
> INFO ::: 2022-12-07 05:37:56 ::: Found 140,239 reference sequences <br/>
> INFO ::: 2022-12-07 05:37:56 ::: Removing references without mappings or less than 10 reads... <br/>
> INFO ::: 2022-12-07 05:37:56 ::: Keeping 22,010 references <br/>
> INFO ::: 2022-12-07 05:37:56 ::: Processing 96 chunks of 230 references each <br/>
> INFO ::: 2022-12-07 05:47:05 ::: Writing reference statistics to GIL-1.vanilla_stats.tsv.gz <br/>
> INFO ::: 2022-12-07 05:47:08 ::: Filtering stats... <br/>
> INFO ::: 2022-12-07 05:47:08 ::: ::: min_read_count >= 10 & min_read_length >= 30 & min_avg_read_ani >= 90.0 & min_expected_breadth_ratio >= 0.5 & min_breadth >= 0 & min_coverage_evenness >= 0 & min_coverage_mean >= 0 <br/>
> INFO ::: 2022-12-07 05:47:08 ::: Saving filtered stats...  <br/>
> INFO ::: 2022-12-07 05:47:10 ::: Writing filtered BAM file... (be patient) <br/>
> INFO ::: 2022-12-07 05:47:11 ::: ::: Filtering 20,858 references sequentially... <br/>
> INFO ::: 2022-12-07 05:47:58 ::: BAM index not found. Indexing... <br/>
> INFO ::: 2022-12-07 05:48:01 ::: ALL DONE. <br/>

* Check how much noise did we filter out:



```
### The number of aligned reads
samtools view -c ../Bowtie2/Lola.vanilla.rmdup.bam
samtools view -c Lola.reassigned.bam
samtools view -c Lola.dedup.filtered.bam
```
> 30251212 <br/>
> 7154105 <br/>
> 7108882 <br/>


```
### The number of reference genomes
tail -n 200 slurm*.out
```

* Check the summary statistic result of filterBAM

```
# overview
zcat GIL-1.hires_stats-filtered.tsv.gz | less -S
# extract the columns we may be interested in
module load csvtk

zcat GIL-1.hires_stats-filtered.tsv.gz \
| csvtk cut -t -T -f "reference,n_reads,read_ani_mean,read_ani_std,coverage_mean,breadth,breadth_exp_ratio,cov_evenness,tax_abund_read" \
| csvtk sort -t -T -k "n_reads:Nr" | less -S
```

3. Assess the damage pattern

Here, we apply the metaDMG to estimate the damage over taxa within samples (LCA mode).

* Load the metaDMG working environment

```
mkdir -p metaDMG
cd metaDMG
cp /projects/course_sgbb20001/people/hsf378/metaDMG/metaDMG.sh ./
conda activate /projects/course_sgbb20001/data/envs/metaDMG/
```
* Generate the config file 

```
 metaDMG config --config-file Lola.vanilla.yaml --custom-database --names /projects/course_sgbb20001/data/databases/Vanilla/names.dmp --nodes /projects/course_sgbb20001/data/databases/Vanilla/nodes.dmp --acc2tax /projects/course_sgbb20001/data/databases/Vanilla/acc2taxid.map.gz --parallel-samples 1 --cores-per-sample 1 --output-dir /projects/course_sgbb20001/people/hsf378/metaDMG --max-position 15 --lca-rank '' ' ' --min-similarity-score 0.95 --max-similarity-score 1 --damage-mode lca --weight-type 1 --metaDMG-cpp /projects/course_sgbb20001/people/tmk528/bin/metaDMG-cpp --forward-only ../filterbam/Lola.dedup.filtered.bam
```

* Compute the results

```
sbatch metaDMG.sh
```

<details><summary>The details on metaDMG.sh</summary>

```
#!/bin/bash

#SBATCH -c 10
#SBATCH --mem 100G
#SBATCH --time 1-00
#SBATCH --reservation=Course_SGBB20001
#SBATCH --account=teaching

metaDMG compute Lola.vanilla.yaml

```

</details>

* Check the results
  1. open a new terminal and run the following code:

```
ssh -L 8050:localhost:8050 XXX@mjolnirhead01fl.unicph.domain
conda activate /projects/course_sgbb20001/data/envs/metaDMG/
metaDMG dashboard --results /PATH TO METADMG RESULTS FOLDER/ --port 8050 --server
```
  2. open your browser

```
http://localhost:8050/
```

